### PR TITLE
feat: restore collection in evm refactor

### PR DESCRIPTION
**Pull request by Lanttcat**
## Description

Closes #MF-819

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/shared/src/hooks/useElementOnScreen.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { useEffect, useRef } from 'react'
+
+export const useElementOnScreen = (callback: (isIntersecting: boolean) => void, option: any) => {
```
Comment (Jack-Works) [inline packages/shared/src/hooks/useElementOnScreen.ts]: I suggest installing `react-hookz`, which is created by the author of `react-use` (react-use is stop maintaining).
<https://react-hookz.github.io/web/?path=/docs/sensor-useintersectionobserver--example>

File: packages/plugins/EVM/src/state/Hub/hub.ts [hunk 2]
```diff
@@ -175,7 +175,7 @@ class Hub implements EVM_Hub {
                 size: this.sizePerPage,
             })
 
-            yield* pageable.data
+            yield yield* pageable.data
```
Comment (Jack-Works) [inline packages/plugins/EVM/src/state/Hub/hub.ts]: Why?

File: packages/shared-base/src/utils/asyncIterator.ts [hunk 3]
```diff
@@ -3,3 +3,7 @@ export async function asyncIteratorToArray<T>(iterable: AsyncIterable<T>) {
     for await (const x of iterable) arr.push(x)
     return arr
 }
+
+export async function* asyncIteratorMerge<T>(iterables: Array<AsyncIterable<T>>) {
```
Comment (Jack-Works) [inline packages/shared-base/src/utils/asyncIterator.ts]:
```suggestion
export async function* flattenAsyncIterator<T>(iterables: Array<AsyncIterable<T>>) {
```

File: packages/dashboard/src/pages/Wallets/components/AddCollectibleDialog/index.tsx [hunk 4]
```diff
@@ -25,48 +34,64 @@ enum FormErrorType {
...
+    const [tokenId, setTokenId] = useState('')
+
+    const { value, loading } = useNonFungibleTokenContract(NetworkPluginID.PLUGIN_EVM, address, {})
+
+    const onSubmit = useCallback(async () => {
+        if (loading || !account || !value || !hub?.getNonFungibleAsset) return
+
+        // If the NonFungible token is added
+        const tokenInDB = trustedNonFungibleTokens.find(
+            (x) => isSameAddress(x.contract?.owner, account) && x.tokenId === tokenId && x.address === address,
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/Wallets/components/AddCollectibleDialog/index.tsx]:
```suggestion
            (x) => isSameAddress(x.contract?.owner, account) && x.tokenId === tokenId && isSameAddress(x.address, address),
```

File: packages/web3-providers/src/alchemy/index.ts [hunk 5]
```diff
@@ -68,6 +68,7 @@ export class Alchemy_EVM_API implements NonFungibleTokenAPI.Provider<ChainId_EVM
 
     getAssets = async (from: string, { chainId = ChainId_EVM.Mainnet, indicator }: HubOptions<ChainId_EVM> = {}) => {
         const chainInfo = Alchemy_EVM_NetworkMap?.chains?.find((chain) => chain.chainId === chainId)
+        if (!chainInfo) return createPageable([], '')
```
Comment (guanbinrui) [inline packages/web3-providers/src/alchemy/index.ts]:
```suggestion
        if (!chainInfo) return createPageable([], createIndicator(options.indicator, ''))
```

File: packages/plugins/Solana/src/state/Hub/hub.ts [hunk 6]
```diff
@@ -126,6 +126,38 @@ class Hub implements SolanaHub {
     ): Promise<Array<Transaction<ChainId, SchemaType>>> {
         throw new Error('Method not implemented.')
     }
+
+    // async *getAllFungibleAssets(address: string): AsyncIterableIterator<FungibleAsset<ChainId, SchemaType>> {
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/state/Hub/hub.ts]: Please remove it.

File: packages/plugins/EVM/src/state/Hub/hub.ts [hunk 7]
```diff
@@ -133,8 +133,15 @@ class Hub implements EVM_Hub {
...
         }
+        const chainProviders: { [key in ChainId]?: Array<typeof OpenSea | typeof Alchemy_EVM | typeof Rarible> } = {
+            [ChainId.Mainnet]: [OpenSea, Alchemy_EVM, Rarible],
+            [ChainId.Matic]: [Alchemy_EVM, OpenSea, Rarible],
+        }
         const predicate = createPredicate(Object.keys(providers) as Array<keyof typeof providers>)
-        const filteredProviders = predicate(sourceType) ? [providers[sourceType]] : [Alchemy_EVM, OpenSea, Rarible]
+        const filteredProviders = predicate(sourceType)
+            ? [providers[sourceType]]
+            : chainProviders[options?.chainId ?? ChainId.Matic] ?? [Alchemy_EVM, OpenSea, Rarible]
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Hub/hub.ts]: Does it better?

```suggestion
            : options?.chainId === ChainId.Matic ? [Alchemy_EVM, OpenSea, Rarible] : [OpenSea, Alchemy_EVM, Rarible]
```

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 8]
```diff
@@ -130,98 +143,77 @@ function CollectibleItem(props: CollectibleItemProps) {
...
-    return null
-    // const { provider, wallet, collectibles, loading, collectiblesRetry, error, readonly, hasRetry = true } = props
-    // const { t } = useI18N()
-    // const classes = useStylesExtends(useStyles(), props)
-
-    // // useEffect(() => WalletMessages.events.erc721TokensUpdated.on(collectiblesRetry))
+    const { provider, wallet, collectibles, loading, collectiblesRetry, error, readonly, hasRetry = true } = props
+    const { t } = useI18N()
+    const classes = useStylesExtends(useStyles(), props)
+    // useEffect(() => WalletMessages.events.erc721TokensUpdated.on(collectiblesRetry))
```
Comment (guanbinrui) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx]:
```suggestion
```

File: packages/web3-providers/src/alchemy/index.ts [hunk 9]
```diff
@@ -68,6 +68,7 @@ export class Alchemy_EVM_API implements NonFungibleTokenAPI.Provider<ChainId_EVM
 
     getAssets = async (from: string, { chainId = ChainId_EVM.Mainnet, indicator }: HubOptions<ChainId_EVM> = {}) => {
         const chainInfo = Alchemy_EVM_NetworkMap?.chains?.find((chain) => chain.chainId === chainId)
+        if (!chainInfo) return createPageable([], createIndicator())
```
Comment (guanbinrui) [inline packages/web3-providers/src/alchemy/index.ts]:
```suggestion
        if (!chainInfo) return createPageable([], createIndicator(indicator))
```

**Pull request merged**